[WONT MERGE] Fix memory leak in subclass and add more error checking

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64551
* #64746
* #64550

Note that this is not a solution for #64548 which should be addressed as well.

This PR adds two main checks:
-  When creating Tensor pyobjects, we ensure that the type is a subtype of THPVariable (otherwise our pointer casts do very weird stuff).
-  Ensures that we only set the Tensor's pyobj if it is not already set (otherwise we leak the THPVariable). Fix call sites where the user explicitly asked for the pyobject to be changed to properly clear it before setting it again.

albanD committed:
Fix memory leak in subclass and add more error checking

[ghstack-poisoned]

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -335,6 +336,34 @@ def backward(ctx, grad_output):
+        class Foo(torch.Tensor):
+            pass
+        a = torch.Tensor._make_subclass(Foo, LoggingTensor(torch.rand(2)))
+        b = LoggingTensor(torch.rand(2)).as_subclass(Foo)
```
Thinking about the preferred behavior here, I kind of want `as_subclass` to hit up `__torch_function__` and the default implementation to raise an error if Foo isn't a subclass of LoggingTensor. Because you have no reason to believe that slicing off all of the LoggingTensor state and turning it into a Foo is a valid operation.

albanD replied:
If we go to `__torch_function__`, how is this default implementation supposed to work? Because even if it is a proper subclass, we will still have the problem that the `.detach()` will generate a Tensor with a pyobj already (of the parent class).
And so if we refuse to "replace" these pyobject, we can't really make such a subclass?

Also, we have some subclass (like nn.Parameter) that disable `__torch_function__` so that might be problematic.

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -335,6 +336,34 @@ def backward(ctx, grad_output):
+        # And in case where we don't know if the user wants this subclass
+        # overwritten, raise a nice error.
+        # The standard LoggingTensor will fail because it is not on the right device
+        with self.assertRaisesRegex(TypeError, "expected.*device=cpu.*device=meta"):
```
TBH, the device error is not very good, we'd prefer the other error

albanD replied:
The problem is that this error is raised by the tensor options check in the legacy constructor before we get to the actual subclass wrapping.
This is true in general for the legacy Tensor constructor:
```python
import torch

a = torch.rand(2, device="gpu")
b = torch.Tensor(a)
```
This will raise this same device error.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -303,9 +346,12 @@ static PyObject* THPVariable_as_subclass(PyObject* _self, PyObject* args, PyObje
     throw torch::TypeError("cls must be a type (got %s)", Py_TYPE(cls)->tp_name);
   }
+  auto data = self.alias();
+  unsafe_clear_pyobj(data);
```
Yeah. So instead of clearing the PyObject, I'd rather just error here (preferably because we attempted to do a `__torch_function__` dispatch and no one said they knew how to do this subclass conversion

albanD replied:
But what about subclasses like `nn.Parameter` for which `__torch_function__` is disabled?

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -105,6 +105,44 @@ class PyInterpreterHolder {
+// This function assumes that it is used just before calling into
+// init_pyobj() (via THPVariable_NewWithVar) in particular for the interpreter
+// state.
+static void unsafe_clear_pyobj(const Variable& tensor) {
```
I object to this function existing. It breaks the invariant that once a C++ tensor is associated with a PyObject, it is always associated with that PyObject

albanD replied:
I agree with that.
I wanted to have a full fix here to be able to drive the discussion.

ezyang reviewed (changes requested):
not very hot on the proposed semantics here

ezyang commented:
@albanD and I discussed this in workchat and Alban is going to prepare a more comprehensive proposal for how to overhaul these semantics (or at least say why the situation is complicated)

albanD committed:
Update on "Fix memory leak in subclass and add more error checking"


Note that this is not a solution for #64548 which should be addressed as well.

This PR adds two main checks:
-  When creating Tensor pyobjects, we ensure that the type is a subtype of THPVariable (otherwise our pointer casts do very weird stuff).
-  Ensures that we only set the Tensor's pyobj if it is not already set (otherwise we leak the THPVariable). Fix call sites where the user explicitly asked for the pyobject to be changed to properly clear it before setting it again.


[ghstack-poisoned]

albanD committed:
Update on "Fix memory leak in subclass and add more error checking"


Note that this is not a solution for #64548 which should be addressed as well.

This PR adds two main checks:
-  When creating Tensor pyobjects, we ensure that the type is a subtype of THPVariable (otherwise our pointer casts do very weird stuff).
-  Ensures that we only set the Tensor's pyobj if it is not already set (otherwise we leak the THPVariable). Fix call sites where the user explicitly asked for the pyobject to be changed to properly clear it before setting it again.


[ghstack-poisoned]
